engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/DelayedKafkaAvroSourceFactory.scala

```diff
@@ -0,0 +1,82 @@
+import pl.touk.nussknacker.engine.kafka.generic.BaseKafkaDelayedSourceFactory
+import pl.touk.nussknacker.engine.kafka.generic.KafkaDelayedSourceFactory._
+import pl.touk.nussknacker.engine.kafka.generic.KafkaTypedSourceFactory.TypeDefinitionParamName
+import pl.touk.nussknacker.engine.kafka.source.KafkaSource
+import pl.touk.nussknacker.engine.kafka.{KafkaConfig, PreparedKafkaTopic, RecordFormatter}
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+
+class DelayedKafkaAvroSourceFactory[K:ClassTag, V:ClassTag](schemaRegistryProvider: SchemaRegistryProvider,
+                                                            processObjectDependencies: ProcessObjectDependencies,
+                                                            timestampAssigner: Option[TimestampWatermarkHandler[ConsumerRecord[K, V]]])
+  extends KafkaAvroSourceFactory[K, V](schemaRegistryProvider, processObjectDependencies, timestampAssigner)
+    with BaseKafkaDelayedSourceFactory {
+
+  override def paramsDeterminedAfterSchema: List[Parameter] = super.paramsDeterminedAfterSchema ++ List(
+    TimestampParameter, DelayParameter
+  )
+
+
+  override protected def nextSteps(context: ValidationContext, dependencies: List[NodeDependencyValue])
+                                  (implicit nodeId: ProcessCompilationError.NodeId): NodeTransformationDefinition = {
+    case step@TransformationStep(
+      (`topicParamName`, DefinedEagerParameter(topic: String, _)) ::
+        (SchemaVersionParamName, DefinedEagerParameter(version: String, _)) ::
+        (TimestampFieldParamName, DefinedEagerParameter(field, _)) ::
+        (DelayParameterName, DefinedEagerParameter(delay, _)) :: Nil, _) =>
+      val preparedTopic = prepareTopic(topic)
+      val versionOption = parseVersionOption(version)
+      val valueValidationResult = determineSchemaAndType(prepareValueSchemaDeterminer(preparedTopic, versionOption), Some(SchemaVersionParamName))
+
+      valueValidationResult match {
+        case Valid((valueRuntimeSchema, _)) =>
+          val definition = valueRuntimeSchema.map(_.schema.getFields.asScala.map(f => (f.name(), ())).toMap ).getOrElse(Map.empty).asJava
+          val delayValidationErrors = Option(delay.asInstanceOf[java.lang.Long]).map(d => validateDelay(d)).getOrElse(Nil)
+          val timestampValidationErrors = Option(field.asInstanceOf[String]).map(f => validateTimestampField(f, definition)).getOrElse(Nil)
+          val errors = delayValidationErrors ++ timestampValidationErrors
+          prepareSourceFinalResults(preparedTopic, valueValidationResult, context, dependencies, step.parameters, errors)
+        case Invalid(exc) =>
+          prepareSourceFinalErrors(context, dependencies, step.parameters, List(exc))
+      }
+    case step@TransformationStep((`topicParamName`, _) :: (SchemaVersionParamName, _) :: (TypeDefinitionParamName, _) :: (TimestampFieldParamName, _) :: (DelayParameterName, _) :: Nil, _) =>
```

Comment:
Why this timestamp field is needed? Until now we have timstamp assignment logic controlled by timestamphandler. Shouldn't be done the same way here? Or maybe timestamp assignment logic should be configurable for all type of kafka sources the same way?

Reply:
We want to allow users to calculate delay related to their selected field that represents "business timestamp".